commit 6fc8150005256db2052b01812d65dff737563a1b
Author: Carlton Gibson <carlton.gibson@noumenal.es>
Date:   Sat Jul 18 14:28:25 2026 +0200

    Fixed #36027 -- Made error response rendering thread-sensitive.
    
    Error response handling frequently makes use of the database, and so
    must use the request specific connection sensitive worker thread.
    
    Previously response_for_exception was dispatched on the event loop's
    default executor. This potentially leads to pool exhaustion, and
    deadlock, when using database pooling option, or a large number of
    permanently held open database connections when not using pooling.
    
    Connections held open were not subject to cleanup. If closed server
    side, the reported "connection is closed" would be seen until
    application restart.
    
    Moving error response rendering to the request specific connection
    sensitive worker thread resolves both of these issues.
    
    The `thread_sensitive=False` was set in
    e17ee4468875077b90b70bb6a589ebad7493f757 to maintain the prior
    behaviour when asgiref switched the thread_sensitive default to
    `True`. There was no specific reason for the non-tread-sensitive
    behaviour beyond that. (With hindsight it was always incorrect for
    the reasons here.)

diff --git a/django/contrib/staticfiles/handlers.py b/django/contrib/staticfiles/handlers.py
index 686718a355..44c1187c8e 100644
--- a/django/contrib/staticfiles/handlers.py
+++ b/django/contrib/staticfiles/handlers.py
@@ -59,7 +59,7 @@ class StaticFilesHandlerMixin:
         try:
             return await sync_to_async(self.serve, thread_sensitive=False)(request)
         except Http404 as e:
-            return await sync_to_async(response_for_exception, thread_sensitive=False)(
+            return await sync_to_async(response_for_exception, thread_sensitive=True)(
                 request, e
             )
 
diff --git a/django/core/handlers/exception.py b/django/core/handlers/exception.py
index 4adc9d7c6a..559291abfc 100644
--- a/django/core/handlers/exception.py
+++ b/django/core/handlers/exception.py
@@ -43,7 +43,7 @@ def convert_exception_to_response(get_response):
                 response = await get_response(request)
             except Exception as exc:
                 response = await sync_to_async(
-                    response_for_exception, thread_sensitive=False
+                    response_for_exception, thread_sensitive=True
                 )(request, exc)
             return response
 
